### ISSUE TITLE
Unused tag applied to too big range

### ISSUE THREAD

**Issue by rchl** [python-lsp/python-lsp-server#253]
In #229 support for `DiagnosticTag` was introduced for flake8 but the range in those diagnostics is not quite right.

With a file like (requires `"pylsp.plugins.flake8.enabled": true`):

```py
from typing import Any, List


def foo() -> Any:
    return 1
```

the whole `from typing` line is marked with "unused" tag while only the `List` import is unused.

Here is how it looks with pylsp:

![Screenshot 2022-08-24 at 08 29 48](https://user-images.githubusercontent.com/153197/186346675-b90418cf-6cd2-4a2a-8bfe-7700ea2e2a87.png)

Here is how it looks with pyright:

![Screenshot 2022-08-24 at 08 30 19](https://user-images.githubusercontent.com/153197/186346745-0638452d-25dd-4a82-b3d4-c31f0a083b04.png)

@krassowski @ccordoba12

**Comment by rchl** [comment]
Diagnostic payload:

```json
{
  "diagnostics": [
    {
      "code": "F401",
      "message": "F401 'typing.List' imported but unused",
      "range": {
        "end": {
          "character": 29,
          "line": 0
        },
        "start": {
          "character": 0,
          "line": 0
        }
      },
      "severity": 1,
      "source": "flake8",
      "tags": [
        1
      ]
    }
  ],
  "uri": "file:///Users/rafal/Documents/testpy/test.py"
}
```

**Comment by rchl** [comment]
Looking at the result, I suppose it might be more flake8 issue that it reports diagnostic for the whole line.

**Comment by krassowski** [comment]
Yes, nothing with support for diagnostic tags, just the old issue of missing precise positions for diagnostics themselves. Note: flake8 is just a wrapper around pyflakes and friends.

Over a year ago I opened a PR to expose the metadata we need (start and end position) upstream https://github.com/PyCQA/pyflakes/pull/649 for an issue which is even older. It was meat with [rather little enthusiasm](https://github.com/PyCQA/pyflakes/issues/454#issuecomment-881906477) by the maintainers at the time, suggesting that they would reconsider once they can drop Python 3.7 (EOL is June 2023) but they still support 3.6 (which is EOL for 8 months now). You can try to nudge/rebase that PR with hope they would reconsider earlier - after all this is not a user facing change (not visible in pyflakes reports).

Otherwise, we could just monkeypatch here (being very defensive - wrapping everything in big try-except) if @ccordoba12 agrees.

**Comment by ccordoba12** [comment]
> Otherwise, we could just monkeypatch here (being very defensive - wrapping everything in big try-except) if @ccordoba12 agrees.

Yeah, I think it'd be very nice to have support for this here because it's confusing for users not being able to see the exact position where an error/warning occurs.

So, I'm quite +1 on this @krassowski.

**Comment by Wuestengecko** [comment]
Hi! I noticed that something very similar applies to the pylint plugin as well, when it marks unused variables and function parameters. It does correctly set the start position in this case, but the highlighting extends until end of line, which can be very confusing:

![grafik](https://user-images.githubusercontent.com/1579756/233789334-1e22187a-29f0-4df0-87f6-7edd0dedf7b8.png)

This one might be easier to fix; considering we already have an accurate start position, we could derive the end position by looking for the end of the "word" – which, if I'm not missing something, should always be an identifier (i.e. `word.isidentifier() is True`).

**Comment by ccordoba12** [comment]
@Wuestengecko, we're very busy at the moment but you're welcome to send us a pull request with a possible fix to take a look at it.
